LINVAST.Imperative/Builders/Java/JavaASTBuilder.Types.cs

```diff
@@ -14,6 +14,193 @@ namespace LINVAST.Imperative.Builders.Java
 {
     public sealed partial class JavaASTBuilder : JavaBaseVisitor<ASTNode>, IASTBuilder<JavaParser>
     {
-        // TODO after types are added to AST
+        public override ASTNode VisitTypeDeclaration([NotNull] TypeDeclarationContext ctx)
+        {
+
+            var ctxStartLine = ctx.Start.Line;
+            var modifiers = "";
+            if (ctx.classOrInterfaceModifier() is { }) {
+                ctxStartLine = ctx.classOrInterfaceModifier().First().Start.Line;
+
+                modifiers = string.Join(" ", ctx.classOrInterfaceModifier().Select(c => c.GetText()));
+
+            }
+
+            if (ctx.annotationTypeDeclaration() is { }) {
+                throw new NotImplementedException("annotations");
+            }
+
+            if (ctx.classDeclaration() is { }) {
+                var classDecl = this.Visit(ctx.classDeclaration()).As<TypeDeclNode>();
+                int declSpecsLine = ctxStartLine;
+                return new ClassNode(ctx.Start.Line, new DeclSpecsNode(declSpecsLine, modifiers, classDecl.Identifier), classDecl);
+            }
+
+            if (ctx.enumDeclaration() is { }) {
+                var enumDecl = this.Visit(ctx.enumDeclaration()).As<EnumDeclNode>();
+                int declSpecsLine = ctxStartLine;
+                return new EnumNode(ctx.Start.Line, new DeclSpecsNode(declSpecsLine, modifiers, enumDecl.Identifier), enumDecl);
+            }
+
+            if (ctx.interfaceDeclaration() is { }) {
+                var interfaceDecl = this.Visit(ctx.interfaceDeclaration()).As<TypeDeclNode>();
+                int declSpecsLine = ctxStartLine;
+                return new InterfaceNode(ctx.Start.Line, new DeclSpecsNode(declSpecsLine, modifiers, interfaceDecl.Identifier), interfaceDecl);
+            }
+            return new EmptyStatNode(ctx.Start.Line);
+        }
+
+        public override ASTNode VisitClassOrInterfaceModifier([NotNull] ClassOrInterfaceModifierContext ctx)
+        {
+
+            if (ctx.annotation() is { })
+                throw new NotImplementedException("annotations");
+
+            return new DeclSpecsNode(ctx.Start.Line, ctx.children.First().GetText());
+
+        }
+
+        public override ASTNode VisitTypeType([NotNull] TypeTypeContext ctx)
+        {
+
+            if (ctx.annotation() is { } && ctx.annotation().Any()) {
+                throw new NotImplementedException("annotations");
+            }
+
+            if (ctx.primitiveType() is { }) {
+                return this.Visit(ctx.primitiveType());
+
+            }
+
+            return this.Visit(ctx.classOrInterfaceType());
+
+        }
+
+
+        public override ASTNode VisitTypeList([NotNull] TypeListContext ctx)
+        {
+            var typeNameNodes = ctx.typeType().Select(c => this.Visit(c).As<TypeNameNode>()).ToArray<TypeNameNode>();
```

Comment:
I don't think that `ToArray` is nececary here? This also applies where-ever the node array is created from an `IEnumerable<T>` and passed to a constructor which takes `IEnumerable<T>`

Reply:
I removed all occurences of _ToArray_.